### ISSUE TITLE
 DynamoDBContext.GetUnconfiguredTable not always the same as AmazonDynamoDBClient.DescribeTable result

### ISSUE THREAD

**Issue by TomFreeman** [aws/aws-sdk-net#602]
Probably a bit of an edge case here, but here's the scenario:

Trying to create an operation that changes the keys on a dynamo table, this seems to be only possible by deleting and recreating the table - if there's a better way then that would be preferable.

This used to work with the old unified SDK.

We're testing using DynamoDB Local (not sure if that's relevant.)

Ignoring that this code is indescribably crude 🙍 

The `Client` property is an instance of `AmazonDynamoDBClient` that is created when the local dynamo is spun up, and retained throughout the test run.

```
private void MoveToDifferentSchema<TFrom, TTo>(Func<TFrom, TTo> transform, List<KeySchemaElement> keySchemaDefinition, List<AttributeDefinition> attributeDefinitions)
{
   TFrom[] existingData;
   using (var context = new DynamoDBContext(Client))
   {
      existingData = context.Scan<TFrom>().ToArray();
   }

   var tableInfo = Task.Run(() => Client.DescribeTableAsync(new DescribeTableRequest()
   {
                TableName = TableUtilities.DetermineTableName<TFrom>()
   }).Result).Result;

   Client.DeleteTable(TableUtilities.DetermineTableName<TFrom>());
   WaitForTableToBeDeleted<TFrom>(TimeSpan.FromSeconds(30));

   var createRequest = new CreateTableRequest(
      TableUtilities.DetermineTableName<TTo>(),
      keySchemaDefinition,
      attributeDefinitions,
      new ProvisionedThroughput(tableInfo.Table.ProvisionedThroughput.ReadCapacityUnits, tableInfo.Table.ProvisionedThroughput.WriteCapacityUnits));

   Client.CreateTable(createRequest);

   WaitForTableToBeCreated<TTo>(TimeSpan.FromSeconds(90));

   BatchWriteItems(existingData.Select(transform));
}

private void BatchWriteItems<T>(IEnumerable<T> data)
{
   using (var context = new DynamoDBContext(Client, new DynamoDBContextConfig()) {
      _logger.Info(GetType(), "Table created, writing migrated data");
      var write = context.CreateBatchWrite<T>();

      write.AddPutItems(data);

      write.Execute();
      _logger.Info(GetType(), "Migrated data written.");
   }
}
```

The code fails on the BatchWriteItems with the error message that "Number of range keys does not match the number of hash keys on table..."

The trigger seems to be that calling `Context.GetUnconfiguredTable` with the tablename returns the previous schema with just one key, whereas the call to `Client.DescribleTable` gives the new Schema.

Given the Context is created new, presumably there's some level of shared cache here around the table configs that isn't up-to-date with the table change.

Anyone able to help?

**Comment by TomFreeman** [comment]
My suspicious appears to be correct, as adding `Table.ClearTableCache();` before doing the write fixes the test.

**Comment by PavelSafronov** [comment]
You are correct, there is an in-process cache and that is exactly what you are running into.
We cover it in some detail in [this blog post](https://aws.amazon.com/blogs/developer/dynamodb-table-cache/), including clearing it out using `Table.ClearTableCache()` as you noticed, or disabling it altogether.

If you are only "modifying" tables during testing, it might be a good idea to disable the cache there.

**Comment by TomFreeman** [comment]
@PavelSafronov Okay, that's fine. Why doesn't the cache reflect operations the SDK itself does?

In this case, I'm deleting a table through the SDK, so surely it should remove that table from the cache?

**Comment by PavelSafronov** [comment]
The approach you describe may work, but would require a number of checks (need to verify that the same account is doing the delete, even if different credentials under a common account are used, in the same region or ServiceUrl), and would only be able to catch a small number of delete scenarios (it would only work if delete is performed in the same process). If we did go with this approach, the solution would be incomplete and fairly complicated.
Additionally as you pointed out this is an edge case, since a large number of users are likely working with unchanging tables, so we decided to go with the simpler approach that will cover almost all of the customers and provided cache controls for the other cases.
